### ISSUE TITLE
Undefined variables (names) are not detected

### ISSUE THREAD

**Issue by axbender** [charliermarsh/ruff#3679]
I would expect the following code to be rejected by ruff  (`select = ["ALL"]`, `ignore = []`):

```python
import pycurl

def have_curl():
    try:
        curl = pycurl.Curl()
    except Exception as e:
        return False
    finally:
        curl.close()   # curl may be undefined here
    return True


def stuff(var):
    try:
        b = 1 / 0
        a = var % 7
    except:
        b = 8
    finally:
        stuff(a)   # a may be undefined here
```
pylint detects these errors. Am I missing something...?

**Comment by axbender** [comment]
Sorry, I just saw that E0601 is on your list for the full pylint implementation...
Will keep waiting hopefully ;-)

**Comment by charliermarsh** [comment]
Yeah we don't have sufficiently sophisticated branch analysis to handle this right now. Going to close just in favor of the broader Pylint issue!
